cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -427,8 +427,38 @@ void ShiftTime(FunctionContext* ctx, const CastOptions& options, const bool is_m
       out_data[i] = static_cast<out_type>(in_data[i]);
     }
   } else if (is_multiply) {
-    for (int64_t i = 0; i < input.length; i++) {
-      out_data[i] = static_cast<out_type>(in_data[i] * factor);
+    if (options.allow_time_truncate) {
+      for (int64_t i = 0; i < input.length; i++) {
+        out_data[i] = static_cast<out_type>(in_data[i] * factor);
+      }
+    } else {
+#define RAISE_OVERFLOW_CAST(VAL)                                                  \
+  ctx->SetStatus(Status::Invalid("Casting from ", input.type->ToString(), " to ", \
+                                 output->type->ToString(), " would result in ",   \
+                                 "out of bounds timestamp: ", VAL));
+
+      if (input.null_count != 0) {
+        internal::BitmapReader bit_reader(input.buffers[0]->data(), input.offset,
+                                          input.length);
+        for (int64_t i = 0; i < input.length; i++) {
+          out_data[i] = static_cast<out_type>(in_data[i] * factor);
+          if (bit_reader.IsSet() && (out_data[i] / factor != in_data[i])) {
```

Comment:
I am dividing again to see if this is then equal the approach used in for truncation when the original divides). But in principle, we could also define minimum and maximum values for each resolution, and compare with those two (that might be more performant?)

Reply:
I think it would be much more performant than doing an integer division on the critical path, yes. You can just compute the min/max dynamically at function start, I think.

Reply:
Updated to compute min/max dynamically